scala-api/src/main/scala/com/cloudera/livy/scalaapi/ScalaJobHandle.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Cloudera, Inc. under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Cloudera, Inc. licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.scalaapi
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{CanAwait, ExecutionContext, Future, TimeoutException}
+import scala.util.Try
+
+import com.cloudera.livy.JobHandle
+import com.cloudera.livy.JobHandle.{Listener, State}
+
+class ScalaJobHandle[T] private[livy] (jobHandle: JobHandle[T]) extends Future[T] {
+
+  private var listener: JobHandle.Listener[T]  = null
+
+  def getState(): State = jobHandle.getState
+
+  override def onComplete[U](func: (Try[T]) => U)(implicit executor: ExecutionContext): Unit = {
+    initiateListener(func, executor)
+    if (isCompleted) {
```

Comment:
This case is already handled by `JobHandle.addListener` (which I don't see you calling).